Investigate the possibility of replacing twisted http server in run-benchmark with python3 http.server

jcarterbohan opened the pull request:
#### 1e3555d85962434a8f26aae28cd16815ab5412b8
<pre>
Investigate the possibility of replacing twisted http server in run-benchmark with python3 http.server
<a href="https://bugs.webkit.org/show_bug.cgi?id=259835">https://bugs.webkit.org/show_bug.cgi?id=259835</a>

Reviewed by Stephanie Lewis.

This is an implementation of the python 3 http.server library as the server hosting benchmark requests. This would possibly replace the implementation where the http server is being run through the externl library, twisted.
Running the http server is behind the new optional flag &quot;--http-server-type&quot; which takes in arguments &quot;twisted&quot; or &quot;base&quot; and will launch the http server when you put in as the argument. If the flag is not used it will default to using the twisted http server.
Requirements for if twisted is installed will only be checked for when running the twisted http server.

* Tools/Scripts/webkitpy/benchmark_runner/README.md:
* Tools/Scripts/webkitpy/benchmark_runner/generic_factory.py:
(GenericFactory.create):
* Tools/Scripts/webkitpy/benchmark_runner/http_server_driver/http_server/builtin_http_server.py: Added.
(Handler):
(Handler.__init__):
(Handler.server_shutdown):
(Handler.log_message):
(Handler.do_GET):
(Handler.do_POST):
* Tools/Scripts/webkitpy/benchmark_runner/http_server_driver/simple_http_server_driver.py:
(SimpleHTTPServerDriver.__init__):
(SimpleHTTPServerDriver.serve):
(SimpleHTTPServerDriver.set_http_server_type):
(SimpleHTTPServerDriver._ensure_http_server_dependencies):
* Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py:
(config_argument_parser):
(run_benchmark_plan):
* Tools/Scripts/webkitpy/benchmark_runner/webserver_benchmark_runner.py:
(WebServerBenchmarkRunner.__init__):
(WebServerBenchmarkRunner):

Canonical link: <a href="https://commits.webkit.org/271176@main">https://commits.webkit.org/271176@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c1b599ffeeee8b669a4058e7987a99a61da0c52b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27546 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6188 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28795 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29771 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25211 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28018 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8111 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3582 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/24993 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27812 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4966 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23641 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4320 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/27706 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4492 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24638 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30410 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25163 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25065 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30615 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4510 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2647 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/28582 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5983 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6631 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4966 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4905 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/http_server_driver/http_server/base_http_server.py:
```diff
@@ -0,0 +1,96 @@
+                shutdown_thread.start()
+                sys.exit(0)
+            except Exception as e:
+                _log.error('Error shutting down HTTP server: %s' % e)
```
Nit: use f-strings in python3 code https://docs.python.org/3/tutorial/inputoutput.html

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/http_server_driver/http_server/base_http_server.py:
```diff
@@ -0,0 +1,96 @@
+                sys.stdout.flush()
+                self.send_response(200)
+            except Exception as e:
+                _log.error('Error reading and writing report: %s' % e)
```
Ditto.

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/http_server_driver/http_server/base_http_server.py:
```diff
@@ -0,0 +1,96 @@
+                shutdown_thread.start()
+                sys.exit(0)
+            except Exception as e:
+                _log.error('Error shutting down HTTP server: %s' % e)
```
Ditto

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/webserver_benchmark_runner.py:
```diff
@@ -23,12 +23,14 @@
         if self._diagnose_dir:
             self._http_server_driver.set_http_log(os.path.join(self._diagnose_dir, 'run-benchmark-http.log'))
+        self._http_server_driver.set_http_server_type(http_server_type)
+        self._http_server_driver._ensure_http_server_dependencies()
```
Hmm, `HTTPServerDriverFactory.create(platform)` calls `_ensure_http_server_dependencies` already,  maybe we should refactor `GenericFactory.create` signature from `create(cls, description)` to
```
def create(cls, description, *args, **kwargs):
    return cls.products[description](*args, **kwargs)
``` 
In this way we can pass in the http server type.

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py:
```diff
@@ -68,6 +68,8 @@ def config_argument_parser():
 
     parser.add_argument('browser_args', nargs='*', help='Additional arguments to pass to the browser process. These are positional arguments and must follow all other arguments. If the pass through arguments begin with a dash, use `--` before the argument list begins.')
 
+    parser.add_argument('--http-server-type', default="twisted", choices=["twisted", "base"], help="Specify the http server to use for the webserver benchmark runner.")
```
`base` does not seem to be accurate enough, how about `builtin` or `py3-builtin`?

jcarterbohan replied:
I'll rename it to `builtin`

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/http_server_driver/http_server/base_http_server.py:
```diff
@@ -0,0 +1,96 @@
+                self.send_response(500)
+            self.end_headers()
+            try:
+                shutdown_thread = threading.Thread(target=self.server.shutdown)
```
Two questions:
1. this seems repeating, may worth a helper function.
2. Curious, is this the recommended way for a graceful shutdown? What will happen if we just do `sys.exit(0)`?

jcarterbohan replied:
Using self.server.shutdown() is the recommended way of shutting down http.server gracefully, this is running on a separate thread in the case there are other incoming requests needing to be handled before the shutdown. 

sys.exit(0) works and could be a little bit faster, but this could possibly cause unexpected behavior in the future.

I'll be adding a helper function to handle the server shutdown.

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/http_server_driver/http_server/base_http_server.py:
```diff
@@ -0,0 +1,96 @@
+import socket
+import threading
+# Since we execute this script directly as a subprocess, we need to ensure
+# that Tools/Scripts is in sys.path for the next imports to work correctly.
```
Do we import anything after this?

jcarterbohan replied:
Nothing is imported after this, this was a leftover from earlier. I'll be taking this part out because it is not needed

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/http_server_driver/http_server/base_http_server.py:
```diff
@@ -0,0 +1,96 @@
+    port = args.port
+    if port == 0:
+        try:
+            port = find_free_port()
```
This is technically subject to race condition.
By digging the source code a bit (https://github.com/python/cpython/blob/3.11/Lib/socketserver.py#L472C26-L472C45 ), I think `HTTPServer((args.interface, port), Handler)` will do exactly what we want: when port == 0, find a free port; when port != 0, try to bind the specified port.

jcarterbohan replied:
If its subject to race condition then I'll let socketserver handle finding a free port

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/http_server_driver/http_server/base_http_server.py:
```diff
@@ -0,0 +1,96 @@
+                sys.exit(1)
+
+        else:
+            super().do_GET()
```
Maybe early return can reduce the indentation:
```
def do_GET(self):
    if not self.path.startswith('/shutdown'):
        super().do_GET()
        return
    ....
```

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/http_server_driver/http_server/base_http_server.py:
```diff
@@ -0,0 +1,96 @@
+                _log.error('Error shutting down HTTP server: %s' % e)
+                sys.exit(1)
+        else:
+            super().do_POST()
```
Ditto: early return style.

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/http_server_driver/http_server/base_http_server.py:
```diff
@@ -0,0 +1,96 @@
+        if self.path.startswith('/report'):
+            length = int(self.headers['content-length'])
+            try:
+                content = self.rfile.read(length)
```
Have we tested if the server will stand while POST a 5MB payload? Or we need a for loop for read by trunk?

jcarterbohan replied:
I tested in this case and it is able to handle 5MB payloads. rfile should be buffering the input by default in the case of larger files.

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/http_server_driver/http_server/base_http_server.py:
```diff
@@ -0,0 +1,96 @@
+            super().do_GET()
+
+    def do_POST(self):
+        if self.path.startswith('/report'):
```
Ideally, we want to do `path = self.translate_path(self.path)` so that `../../report` can hit this endpoint.

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/http_server_driver/http_server/base_http_server.py:
```diff
@@ -0,0 +1,96 @@
+        _log.info(format % args)
+
+    def do_GET(self):
+        if self.path.startswith('/shutdown'):
```
Ditto: `path = self.translate_path(self.path)`

sundiamonde reviewed (commented):
Please rerun EWS before landing and make sure the webkitpy tests are in good state

jcarterbohan committed:
Investigate the possibility of replacing twisted http server in run-benchmark with python3 http.server
https://bugs.webkit.org/show_bug.cgi?id=259835

Reviewed by Stephanie Lewis.

This is an implementation of the python 3 http.server library as the server hosting benchmark requests. This would possibly replace the implementation where the http server is being run through the externl library, twisted.
Running the http server is behind the new optional flag "--http-server-type" which takes in arguments "twisted" or "base" and will launch the http server when you put in as the argument. If the flag is not used it will default to using the twisted http server.
Requirements for if twisted is installed will only be checked for when running the twisted http server.

* Tools/Scripts/webkitpy/benchmark_runner/README.md:
* Tools/Scripts/webkitpy/benchmark_runner/generic_factory.py:
(GenericFactory.create):
* Tools/Scripts/webkitpy/benchmark_runner/http_server_driver/http_server/builtin_http_server.py: Added.
(Handler):
(Handler.__init__):
(Handler.server_shutdown):
(Handler.log_message):
(Handler.do_GET):
(Handler.do_POST):
* Tools/Scripts/webkitpy/benchmark_runner/http_server_driver/simple_http_server_driver.py:
(SimpleHTTPServerDriver.__init__):
(SimpleHTTPServerDriver.serve):
(SimpleHTTPServerDriver.set_http_server_type):
(SimpleHTTPServerDriver._ensure_http_server_dependencies):
* Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py:
(config_argument_parser):
(run_benchmark_plan):
* Tools/Scripts/webkitpy/benchmark_runner/webserver_benchmark_runner.py:
(WebServerBenchmarkRunner.__init__):
(WebServerBenchmarkRunner):

Canonical link: https://commits.webkit.org/271176@main
